runtime/src/consensus/tendermint/merkle.rs

```diff
@@ -0,0 +1,173 @@
+//!
+//! Helper functions copied from:
+//! https://github.com/informalsystems/tendermint-rs/blob/main/tendermint/src/merkle.rs
+
+use std::cmp::Ordering;
+
+use anyhow::{anyhow, Result};
+use rustc_hex::ToHex;
+use sha2::{Digest, Sha256};
+
+use tendermint::merkle::{Hash, HASH_SIZE};
+
+/// Proof represents a Merkle proof.
+///
+/// NOTE: The convention for proofs is to include leaf hashes but to
+/// exclude the root hash.
+/// This convention is implemented across IAVL range proofs as well.
+/// Keep this consistent unless there's a very good reason to change
+/// everything.  This also affects the generalized proof system as
+/// well.
+#[derive(Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct Proof {
+    pub total: i64,       // Total number of items.
+    pub index: i64,       // Index of item to prove.
+    pub leaf_hash: Hash,  // Hash of item value.
+    pub aunts: Vec<Hash>, // Hashes from leaf's sibling to a root's child.
+}
+
+impl Proof {
+    /// Verify that the Proof proves the root hash.
+    /// Check index/total manually if needed.
+    pub fn verify(&self, root_hash: Hash, leaf: Hash) -> Result<()> {
+        if self.total < 0 {
+            return Err(anyhow!("proof total must be positive"));
+        }
+        if self.index < 0 {
+            return Err(anyhow!("proof index cannot be negative"));
+        }
+
+        let leaf_hash = leaf_hash(&leaf);
+        match self.leaf_hash.cmp(&leaf_hash) {
+            Ordering::Equal => (),
+            _ => {
+                return Err(anyhow!(
+                    "invalid leaf hash: wanted {} got {}",
+                    leaf_hash.to_hex::<String>(),
+                    self.leaf_hash.to_hex::<String>(),
+                ));
+            }
+        }
+
+        let computed_hash = self.compute_root_hash().ok_or_else(|| {
+            anyhow!(
+                "invalid root hash: wanted {} got None",
+                root_hash.to_hex::<String>()
+            )
+        })?;
+        match computed_hash.cmp(&root_hash) {
+            Ordering::Equal => (),
+            _ => {
+                return Err(anyhow!(
+                    "invalid root hash: wanted {} got {}",
+                    root_hash.to_hex::<String>(),
+                    computed_hash.to_hex::<String>(),
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Compute the root hash given a leaf hash. Does not verify the result.
+    pub fn compute_root_hash(&self) -> Option<Hash> {
+        Proof::compute_hash_from_aunts(self.index, self.total, self.leaf_hash, &self.aunts)
+    }
+
+    /// Use the leaf_hash and inner_hashes to get the root merkle hash.
+    /// If the length of the inner_hashes slice isn't exactly correct, the result is None.
+    /// Recursive impl.
+    fn compute_hash_from_aunts(
+        index: i64,
+        total: i64,
+        leaf_hash: Hash,
+        inner_hashes: &Vec<Hash>,
```

Comment:
Could this be changed to `&[Hash]` to avoid conversions to vec?

Reply:
Yes, much better.